solr/core/src/java/org/apache/solr/servlet/RequestRateLimiter.java

```diff
@@ -0,0 +1,199 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.servlet;
+
+import javax.servlet.AsyncContext;
+import javax.servlet.AsyncEvent;
+import javax.servlet.AsyncListener;
+import javax.servlet.FilterConfig;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.lang.invoke.MethodHandles;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Handles rate limiting for a specific request type.
+ *
+ * The control flow is as follows:
+ * Handle request -- Check if slot is available -- If available, acquire slot and proceed --
+ * else asynchronously queue the request.
+ *
+ * When an active request completes, a check is performed to see if there are any pending requests.
+ * If there is an available pending request, process the same.
+ */
+public class RequestRateLimiter {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  private Semaphore allowedConcurrentRequests;
+  private RateLimiterConfig rateLimiterConfig;
+  private Queue<AsyncContext> waitQueue;
+  private Queue<AsyncListener> listenerQueue;
+
+  public RequestRateLimiter(RateLimiterConfig rateLimiterConfig) {
+    this.rateLimiterConfig = rateLimiterConfig;
+    this.allowedConcurrentRequests = new Semaphore(rateLimiterConfig.allowedRequests);
+    this.waitQueue = new ConcurrentLinkedQueue<>();
+    this.listenerQueue = new ConcurrentLinkedQueue<>();
```

Comment:
We never poll this queue, what is it used for?

Reply:
This exists to store the listeners that are registered for the given async contexts. Removed the same, thanks